src/v/serde/serde.h

```diff
@@ -0,0 +1,260 @@
+    template<typename T>                                                       \
+    inline constexpr auto const SERDE_TOKEN_PASTE(                             \
+      has_, name) = SERDE_TOKEN_PASTE(help_has_, name)<T>::value;
+
+static auto iob = iobuf{};
+static auto iop = iobuf_parser{std::move(iob)};
+
+HAS_MEMBER_FN(serde_read, iop, version_t{0}, version_t{0}, size_t{0U})
+HAS_MEMBER_FN(serde_write, iob)
+HAS_MEMBER_FN(serde_async_read, iop, version_t{0}, version_t{0}, size_t{0U})
+HAS_MEMBER_FN(serde_async_write, iob)
+
+template<typename T>
+inline constexpr auto const is_serde_compatible_v =
+  is_envelope_v<T> || std::is_scalar_v<T> || std::is_enum_v<T> ||
+  reflection::is_std_vector_v<T> || reflection::is_named_type_v<T> ||
+  reflection::is_ss_bool_v<T> || std::is_same_v<T, std::chrono::milliseconds> ||
+  std::is_same_v<T, iobuf> || std::is_same_v<T, ss::sstring> ||
+  reflection::is_std_optional_v<T>;
+
+using header_t = std::tuple<version_t, version_t, size_t>;
+
+void write_envelope() {}
+
+template<typename T>
+void write(iobuf& out, T const& t) {
+    using Type = std::decay_t<T>;
+    static_assert(has_serde_write<Type> || is_serde_compatible_v<Type>);
+
+    if constexpr (is_envelope_v<Type>) {
+        write(out, Type::__version);
+        write(out, Type::__compat_version);
+
+        auto size_placeholder = out.reserve(
+          sizeof(typename Type::envelope_size_t));
+        auto const size_before = out.size_bytes();
+
+        if constexpr (has_serde_write<Type>) {
+            t.serde_write(out);
+        } else {
+            envelope_for_each_field(t, [&out](auto& f) { write(out, f); });
+        }
+
+        auto const written_size = out.size_bytes() - size_before;
+        if (
+          written_size >
+          std::numeric_limits<typename Type::envelope_size_t>::max()) {
+            throw std::system_error{
+              make_error_code(error_codes::envelope_too_big)};
+        }
+        auto const size = ss::cpu_to_le(
+          static_cast<typename Type::envelope_size_t>(written_size));
+        size_placeholder.write(
+          reinterpret_cast<char const*>(&size),
+          sizeof(typename Type::envelope_size_t));
+    } else if constexpr (std::is_enum_v<Type>) {
+        write(out, static_cast<std::underlying_type_t<Type>>(t));
+    } else if constexpr (std::is_scalar_v<Type>) {
+        auto le_t = ss::cpu_to_le(t);
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(Type));
+    } else if constexpr (reflection::is_std_vector_v<Type>) {
+        assert(t.size() <= std::numeric_limits<int32_t>::max());
+        write(out, static_cast<int32_t>(t.size()));
+        for (auto const& el : t) {
+            write(out, el);
+        }
+    } else if constexpr (reflection::is_named_type_v<Type>) {
+        return write(iob, static_cast<typename Type::type>(t));
+    } else if constexpr (reflection::is_ss_bool_v<Type>) {
```

Comment:
if there are only a couple `reflection::*` things we might consider importing them into `serde::*` since that seems like it is probably the future and then we'll have less coupling.

Reply:
lets' wait